### ISSUE TITLE
Clean up BigPipe documentation, investigate mod_proxy_fcgi further

### ISSUE THREAD

**Issue by geerlingguy** [geerlingguy/drupal-vm#532]
Two things:

Wim Leers mentioned that we shouldn't need to disable output buffering at all in PHP, since BigPipe does the output flushing correctly. So one less thing to configure!

Secondly, maybe there's a way to get things working via `mod_proxy_fcgi`. Just a lot of vague references like this in my initial investigation:

  - http://stackoverflow.com/questions/33628918/disabling-output-buffer-with-apache-and-php-fpm-via-mod-proxy
  - http://stackoverflow.com/questions/30707792/how-to-disable-buffering-with-apache2-and-mod-proxy-fcgi

**Comment by geerlingguy** [comment]
So... looks like BigPipe _does_ work with `mod_proxy_fcgi` (woohoo!), but the output buffers a minimum of 4096 characters (or somewhere thereabouts), and even if I try adding `flushpackets=[on|auto]` and/or `iobuffersize=512` (the minimum allowed, which overrides [ProxyIOBufferSize](https://httpd.apache.org/docs/2.4/mod/mod_proxy.html#proxyiobuffersize) for the worker), the buffer seems to require at least ~4096 bytes of data before the first chunk of data is passed through after a `flush()` from PHP (see the [ProxyPass](https://httpd.apache.org/docs/2.4/mod/mod_proxy.html#proxypass) documentation for all the key/value descriptions).

So, I need to update the documentation (both here and on Drupal.org), stating that BigPipe should work _out of the box_ with Apache, but that in rare circumstances (e.g. you have a teensy tiny site template with < 4096 characters of data in the general markup), Apache could buffer some of the data.

This updated test script is helpful for debugging exactly where the chunking happens—play around with different `$string_lengths` and see what happens:

```php
<?php
// Set a valid header so browsers pick it up correctly.
header('Content-type: text/html; charset=utf-8');

// Emulate the header BigPipe sends so we can test through Varnish.
header('Surrogate-Control: BigPipe/1.0');

// Explicitly disable caching so Varnish and other upstreams won't cache.
header("Cache-Control: no-cache, must-revalidate");

// Setting this header instructs Nginx to disable fastcgi_buffering and disable
// gzip for this request.
header('X-Accel-Buffering: no');

$string_length = 4096;
echo 'Begin test with an ' . $string_length . ' character string...<br />' . "\r\n";

// For 3 seconds, repeat the string.
for ($i = 0; $i < 3; $i++) {
  // 1024 byte string.
  $one_kilobyte_string = str_repeat('.', $string_length);
  echo $one_kilobyte_string . '<br />' . "\r\n";
  echo $i . '<br />' . "\r\n";
  flush();
  sleep(1);
}

echo 'End test.<br />' . "\r\n";
?>
```

**Comment by geerlingguy** [comment]
One more note—you also have to disable `mod_deflate` (`sudo a2dismod deflate` in Ubuntu), otherwise Apache will wait until the end of the response and gzip the output.

Unlike Nginx, there's no simple way to dynamically disable gzip, so you either have to disable the module entirely, or you can (in Apache vhost config, or even through .htaccess) use `SetEnv no-gzip 1` (see: http://serverfault.com/a/435835/15673).

So I'd still highly recommend using Nginx, but at least things should work (mostly) out of the box with Apache.

**Comment by geerlingguy** [comment]
More investigation—the output buffer is actually based on PHP's output_buffering setting—so the default is 4096 (thus the note above), and you can change `php_output_buffering: 512` if you want to lower the buffer to 512 bytes.

**Comment by geerlingguy** [comment]
I also wanted to spend a bit more time confirming whether Gzipping had to be disabled in Apache in all circumstances to get streaming to work. Here are my test results:

## `mod_php`

Apache's `mod_php` seems to be able to handle streaming without disabling deflate/gzip for requests.

## `mod_fastcgi`

I installed `mod_fastcgi` by doing the following:

  1. `
  2. Make `/etc/apache2/mods-enabled/fastcgi.conf` look like the following:

      ```
      <IfModule mod_fastcgi.c>
        AddHandler fastcgi-script .fcgi
        #FastCgiWrapper /usr/lib/apache2/suexec
        FastCgiIpcDir /var/lib/apache2/fastcgi
        AddHandler php5-fcgi .php
        Action php5-fcgi /php5-fcgi
        Alias /php5-fcgi /usr/lib/cgi-bin/php5-fcgi
        FastCgiExternalServer /usr/lib/cgi-bin/php5-fcgi -host 127.0.0.1:9000 -pass-header Authorization
        <Directory /usr/lib/cgi-bin>
          Require all granted
        </Directory>
      </IfModule>
      ```

`mod_fastcgi` requires deflate/gzip to be disabled, else the full response is buffered.

## `mod_proxy_fcgi`

`mod_proxy_fcgi` requires deflate/gzip to be disabled, else the full response is buffered.

**Comment by oxyc** [comment]
Absolutely awesome! Looking forward to trying this out.

**Comment by geerlingguy** [comment]
Blog post on the topic: http://www.jeffgeerling.com/blog/2016/streaming-php-disabling-output-buffering-php-apache-nginx-and-varnish

**Comment by galapogos01** [comment]
Hoping for a little help as this is the most extensive resource I can find on buffering with proxy fcgi!

Running Debian buster with apache 2.4.38, php7.3-fpm and mod_proxy_fcgi. I have tried this on two machines.

I have got this in my apache php conf

    <FilesMatch ".+\.ph(ar|p|tml)$">
        SetHandler "proxy:unix:/run/php/php7.3-fpm.sock|fcgi://localhost/"
    </FilesMatch>

<Proxy "fcgi://localhost/" enablereuse=on flushpackets=on max=10>
</Proxy>
SetEnv no-gzip 1

I can't for the life of me get it to flush the buffer. My headers show gzip is disabled. Even disabling output buffering in the php ini makes no difference. Run the script at the console and it buffers so it's definitely an apache<>php-fpm issue.

Any tips appreciated!
